Add explicit sub-canopy 3-layers effect in photolysis and gas-phase (…

iri01 opened the pull request:
…cb6r3_ae6_aq mechanism)

activated with CANOPY_SHADE namelist parameter.

<!--Instructions: All subsequent sections of text should be filled in as appropriate.-->
## PR Checklist
- [x] This PR has been tested on an RDHPCS machine and/or WCOSS2. Please select below:
  - [x] RDHPCS.
  - [ ] WCOSS2.

- [x] This PR has been tested with the ufs-srweather-app workflow online-cmaq branch.

- [x] New or updated input data is required by this PR. <!-- If checked, please work with the code managers to update input data sets on all platforms.-->

- [x] Baselines are expected to change.

## Description
<!--Provide a detailed description of what this PR does.-->

## Issue(s) addressed
<!--Please link the issues to be closed with this PR.
(Remember, issues must always be created before starting work on a PR branch!)
EX: - fixes #<issue_number>-->

## Dependencies
<!--If testing this branch requires non-default branches in other repositories, list them. Those branches should have matching names (ideally).
Do PRs in upstream/other repositories need to be merged along with this one?
EX: - waiting on noaa-emc/fv3atm/pull/<pr_number>-->

iri01 committed:
Add explicit 3-sub-layers canopy effect in photolysis and gas-phase (cb6r3_ae6_aq mechanism)
activated with CANOPY_SHADE namelist parameter.

drnimbusrain commented on src/model/src/can_mask.F90:
```diff
@@ -0,0 +1,190 @@
+
+      !NOT a Continuos forest canopy
+      IF (                     Met_Data%LAIE(COL,ROW)  .LT. 0.1     &
+           .OR.                Met_Data%FCH (COL,ROW)  .LT. 0.5     &
```
I thought in your recent presentations we are using a FCH threshold of 3.0 meters?

iri01 replied:
As discussed, the FCH threshold I used in my runs is 0.5 m both in photolysis and diffusion.

drnimbusrain commented on src/model/src/can_mask.F90:
```diff
@@ -0,0 +1,190 @@
+
+      !NOT a Continuos forest canopy
+      IF (                     Met_Data%LAIE(COL,ROW)  .LT. 0.1     &
+           .OR.                Met_Data%FCH (COL,ROW)  .LT. 0.5     &
```
I thought in your recent presentations we are using a FCH threshold of 3.0 meters?

drnimbusrain reviewed (changes requested):
@iri01 

Looks OK as I go through it this morning. I think after retroactively, at our next meeting we go through these changes.  I only have a small comment on the canopy mask and FCH threshold here compared to the recent results you have shown.  

I do have one major question here.   Are we missing your _hrdriver.F_ changes here where the gas phase solver is applied on canopy levels? If so please add the required changes (in make files, src/model/src/hrdriver.F files, etc.) here.

 Also, thinking ahead, we have moved to using the Rosenbrock (RB) solver instead of EBI in future AQMv8.  Therefore, I think if possible you should add the corresponding gas phase solver/driver changes for RB to this PR (e.g., https://github.com/USEPA/CMAQ/blob/5.2.1/CCTM/src/gas/ros3/rbdriver.F#L638), so that when we merge in CMAQ54 updates we get the canopy changes for AQMv8 RB solver option. 

Thank you!

iri01 committed:
Add explicit canopy effect on 3 sub layers in CB6 gas-phase mechanism  ("hrdriver")

iri01 committed:
Add canopy modifications to the CB6 gas-phase routines to local codes.

iri01 committed:
Add 3-layer sub-canopy effect to Rosenbrock solver

drnimbusrain reviewed (commented):
@iri01 Thanks looks good from my perspective here.  Just need a short test of this package (i.e., [feature/aqm_canopy_phot3lev](https://github.com/noaa-oar-arl/AQM/tree/feature/aqm_canopy_phot3lev)) with CMAQ521 still using both the HR (EBI) and RB solvers with canopy on - off.  Then we can merge!  Can you test this package with both solvers to make sure they work and get results as expected?   
You have already updated and added local versions of both HR and RB here and to the cmake (for canopy), and just need to make one small change to your cmake path locally to test the RB solver, as Wei has done here:  https://github.com/noaa-oar-arl/AQM/commit/b33ded4fde3d4d2b904c1906545a18417b8566c3#diff-b370bb911efd47962ec736d4070a44da7a643b21dca9e4a90ea087b1910eb4c6R79

iri01 committed:
Add explicit canopy effect (on 3 sub layers) using Rosenbrock solver ("cb6r3_ae6_aq" gas-phase mechanism)
Update explicit canopy effect (on 3 sub layers) using EBI solver -- for consistency with Rosenbrock solver ("cb6r3_ae6_aq" gas-phase mechanism)

iri01 committed:
Definitions of canopy layers moved out of can_mask and made accessible to CB6 solvers directly (hrdriver/rbdriver)

iri01 committed:
Update to include the canopy layers definitions

iri01 committed:
Clean unused canopy block size arrays

drnimbusrain commented on src/model/src/GRID_CONF.F:
```diff
@@ -0,0 +1,110 @@
+C parameters.
+C VGRD_DEFN replaces COORD.EXT and VGRD.EXT
+C Revision History: Jeff Young 31 Jen 05: created
+C.......................................................................
```
@iri01  Please just a small comment with date  here about what you changed for our local AQM version of GRID_CONF with canopy.

iri01 replied:
This a CCTM local version of GRID_CONF (copy from CCTM/grid/cartesian/GRID_CONF.F) with added sub-canopy layers parameters. Dated on January 20, 2025. See lines:

      INTEGER, PARAMETER :: NLAYC = 3 ! # of canopy layers for shading effects
      INTEGER, SAVE :: NLAYT
      INTEGER, SAVE :: MXCELLS_CAN
      INTEGER, SAVE :: MXBLKS_CAN
      ...
      NLAYT = NLAYS + NLAYC
      WRITE( LOGDEV, *) 'GRID_INIT: CANOPY LAYERS = ', NLAYC, NLAYT
      MXCELLS_CAN = NCOLS * NROWS * NLAYT
      MXBLKS_CAN  = 1 + ( MXCELLS_CAN - 1 ) / BLKSIZE
      WRITE( LOGDEV, *) 'GRID_INIT: MXBLKS_CAN = ', MXBLKS_CAN, MXCELLS_CAN,BLKSIZE

drnimbusrain replied:
Great, please add comments into the code here and below

drnimbusrain commented on src/model/src/can_levs_defn.F90:
```diff
@@ -0,0 +1,927 @@
+   module can_levs_defn
+
```
@iri01 Just for now and future updates, a brief date and description of what this new module does (and briefly what subroutines are included) for everyone's record.

drnimbusrain commented on src/model/src/can_mask.F90:
```diff
@@ -0,0 +1,179 @@
+   module can_mask
+
```
@iri01 Just for now and future updates, a brief date and description of what this new module does (and briefly what subroutines are included) for everyone's record.

iri01 replied:
Date: 2024/04. Description: A new canopy module that defines a mask of the canopy shaded columns based on various criteria for 5 canopy datasets. The routine is used in the canopy routines and the gas-phase driver routines (EBI/ROS3)

drnimbusrain commented on src/model/src/can_trans_mod.F90:
```diff
@@ -0,0 +1,822 @@
+module can_trans_mod
+
```
@iri01 Just for now and future updates, a brief date and description of what this new module does (and briefly what subroutines are included) for everyone's record.

drnimbusrain commented on src/model/src/hrdata_mod.F:
```diff
@@ -0,0 +1,289 @@
+C  REVISION HISTORY: Created by EBI solver program, Mar 10, 2017
+C
+C   18 Jul 14 B.Hutzell: revised to use real(8) variables for solution
+C*************************************************************************
```
Small comment on your updates and date.

iri01 replied:
Date: 2024/06. Modified version of EBI gas-phase solver where the gas-phase chemistry (CB6 and ROS3 mechanisms) is computed over canopy-shaded regions, as defined in "canopy_mask" routine, on 3 sub-canopy layers. Diagnostics for gas-phase tendencies and other auxiliary diagnostics are also added.

drnimbusrain reviewed (changes requested):
Looks OK to me, great work in getting this in!  I think maybe only a few comments at the top of the new canopy routines added, briefly explaining what they do starting a change log for tracking changes.

iri01 commented on src/model/src/can_levs_defn.F90:
Date: 2024/06. Description: Initialization of meteorology vars (temperature, humidity, pressure etc.) on 3 canopy sub-layers, which are interpolated between the 1st hybrid model layer and 2m diagnostics from the Land-Surface Model. Used in canopy routines and in the gas-phase chemistry routines.

iri01 commented on src/model/src/can_mask.F90:
Date: 2024/04. Description: A new canopy module defines a mask of the canopy shaded columns based on various criteria for 5 canopy datasets. Used in other canopy routines and the gas-phase driver routines (EBI/ROS3)

drnimbusrain commented:
@iri01 These comments are great, but can you please add them as commits directly in the top of these respective new canopy codes/routines in your branch?

iri01 commented on src/model/src/can_trans_mod.F90:
Date: 2024/09. Description: New canopy routine that transfers mass of chemical tracers between the canopy layers and the model layers, and conserves mass.

iri01 commented on src/model/src/hrdata_mod.F:
Date: 2024/06. Modified version of EBI gas-phase solver where the gas-phase chemistry (CB6 and ROS3 mechanisms) is computed over canopy-shaded regions, as defined in "canopy_mask" routine, on 3 sub-canopy layers. Diagnostics for gas-phase tendencies and other auxiliary diagnostics are also added.

iri01 commented on src/model/src/hrdriver.F:
Date: 2024/06. Modified version of EBI gas-phase solver where the gas-phase chemistry (CB6 and ROS3 mechanisms) is computed over canopy-shaded regions, as defined in "canopy_mask" routine, on 3 sub-canopy layers. Diagnostics for gas-phase tendencies and other auxiliary diagnostics are also added.

iri01 commented on src/model/src/phot.F:
Date: 2024/06. Description: Modified version of photolysis routine, where the photolysis rates over canopy-shaded regions are calculated on 3 sub-canopy layers. Activate photolysis diagnostics. The routine is used in the gas-phase chemistry solvers (EBI/ROS3) routines.

iri01 committed:
Add description to the routines modified with the canopy effects
